[Vectorization] Add polybench benchmarks

JuniMay opened the pull request:
This PR adds Polybench to vectorization benchmarks.

Most of the Polybench cases uses parametric loop bounds and sizes, which makes it easy to run the benchmark with different dataset sizes. But there are some that uses compile-time constant sizes (including fix-sized local arrays). The MLIR source code here are generated by [Polygeist](https://github.com/llvm/Polygeist) and modified manually to run on different dataset sizes dynamically.

The binary `vectorization-polybench-benchmark` provides two options, `run-benchmark` and `run-validation`.

1. `run-benchmark` runs the benchmark with the specified sizes and prints the execution time of each benchmark. To disable certain dataset sizes, use the `--disable-sizes` option. For example, to disable the mini, large, and extralarge dataset sizes, use the following command:

    ```bash
    ./bin/vectorization-polybench-benchmark run-benchmark --disable-sizes=mini,large,extralarge
    ```

2. `run-validation` runs the benchmark with the specified dataset size and output the result to stdout. The format of output data is the same as the original Polybench suite. So this output can be used to validate the correctness of this port.

    ```bash
    ./bin/vectorization-polybench-benchmark run-validation mini
    ```

    Note that the output can be very large, so it is recommended to redirect the output to a file and compare it with the original Polybench output later.

A benchmarking output on RISC-V platform would be like:

```
Disabled dataset sizes: [ extralarge large mini ]
2025-01-14T19:22:37+08:00
Running ./bin/vectorization-polybench-benchmark
Run on (8 X 1600 MHz CPU s)
CPU Caches:
  L1 Instruction 32 KiB (x8)
  L1 Data 32 KiB (x8)
  L2 Unified 512 KiB (x2)
Load Average: 2.35, 2.09, 2.04
----------------------------------------------------------------
Benchmark                      Time             CPU   Iterations
----------------------------------------------------------------
2mm-small                   2.41 ms         2.41 ms          291
2mm-medium                   186 ms          186 ms            4
3mm-small                   3.78 ms         3.78 ms          185
3mm-medium                   371 ms          371 ms            2
adi-small                   13.8 ms         13.8 ms           51
adi-medium                   530 ms          530 ms            1
atax-small                 0.080 ms        0.080 ms         8720
atax-medium                0.902 ms        0.906 ms          772
bicg-small                 0.209 ms        0.209 ms         3346
bicg-medium                 2.38 ms         2.38 ms          294
cholesky-small              2.08 ms         2.08 ms          337
cholesky-medium              112 ms          112 ms            6
correlation-small           2.55 ms         2.55 ms          273
correlation-medium           117 ms          117 ms            6
covariance-small            2.51 ms         2.51 ms          276
covariance-medium            117 ms          117 ms            6
deriche-small               1.42 ms         1.42 ms          493
deriche-medium              23.2 ms         23.2 ms           30
doitgen-small               2.76 ms         2.76 ms          254
doitgen-medium              45.2 ms         45.2 ms           16
durbin-small               0.107 ms        0.107 ms         6508
durbin-medium               1.14 ms         1.14 ms          615
fdtd-2d-small               2.57 ms         2.57 ms          272
fdtd-2d-medium              74.7 ms         74.6 ms            9
floyd-warshall-small        30.7 ms         30.7 ms           23
floyd-warshall-medium        686 ms          686 ms            1
gemm-small                  1.01 ms         1.01 ms          691
gemm-medium                 25.8 ms         25.8 ms           27
gemver-small               0.234 ms        0.234 ms         2995
gemver-medium               9.26 ms         9.26 ms           75
gesummv-small              0.121 ms        0.121 ms         5793
gesummv-medium             0.920 ms        0.923 ms          758
gramschmidt-small           2.99 ms         2.99 ms          234
gramschmidt-medium           154 ms          154 ms            5
heat-3d-small               6.56 ms         6.56 ms          107
heat-3d-medium               241 ms          241 ms            3
jacobi-1d-small            0.028 ms        0.028 ms        25294
jacobi-1d-medium           0.203 ms        0.203 ms         3445
jacobi-2d-small             3.70 ms         3.69 ms          189
jacobi-2d-medium             198 ms          198 ms            4
lu-small                    4.02 ms         4.01 ms          174
lu-medium                    673 ms          673 ms            1
ludcmp-small                3.86 ms         3.86 ms          181
ludcmp-medium                642 ms          641 ms            1
mvt-small                  0.147 ms        0.147 ms         4766
mvt-medium                  8.15 ms         8.15 ms           86
nussinov-small              4.92 ms         4.92 ms          142
nussinov-medium              253 ms          253 ms            3
seidel-2d-small             17.2 ms         17.2 ms           41
seidel-2d-medium             520 ms          520 ms            1
symm-small                  2.44 ms         2.44 ms          287
symm-medium                  113 ms          113 ms            6
syr2k-small                 3.48 ms         3.48 ms          201
syr2k-medium                 119 ms          119 ms            6
syrk-small                  2.08 ms         2.08 ms          336
syrk-medium                 63.3 ms         63.3 ms           11
trisolv-small              0.054 ms        0.054 ms        12922
trisolv-medium             0.830 ms        0.834 ms          838
trmm-small                 0.948 ms        0.949 ms          737
trmm-medium                 51.9 ms         51.9 ms           13
```

zhanghb97 reviewed (commented):
Nice PR!  
Giving some quick feedback about the format:  
- Missing license header in (.h/.cpp) files. Refer to other files in the repository for guidance.  
- Please include additional comments in the (.h/.cpp) files. The current version is not very user-friendly for extensions and is challenging for reviewers to assess.
- Certain MLIR files contain empty lines at the start (e.g., MLIRPolybenchBicg.mlir, etc.).

I need more time to review this PR. Also, ping @xlinsist for review.

JuniMay commented:
I refactored the benchmark registration methods to reuse the google benchmark framework as much as possible and added `-O3` kernels to the benchmarks. Now the benchmarks can be run with the following command:

```bash
./bin/vectorization-polybench-benchmark --benchmark_filter=".*/.*/[12]" --verification-dataset-size=mini
```

which runs all Polybench cases on `small` and `medium` dataset sizes (corresponding to regex `[12]`), and then verify the results of optimized version on `mini` dataset size.

The filter can also be used to run only certain Polybench cases.

Additionally, a simple python script is added to help verify the correctness of this port. Usages of it are placed under the _Verifying the Benchmark Results_ section in README.

A sample of new benchmarking output on RISC-V platform:

```bash
$ ./bin/vectorization-polybench-benchmark --benchmark_filter="MLIRPolybench2mm/.*/[012]" --verification-dataset-size=mini
2025-01-18T21:20:06+08:00
Running ./bin/vectorization-polybench-benchmark
Run on (8 X 1600 MHz CPU s)
CPU Caches:
  L1 Instruction 32 KiB (x8)
  L1 Data 32 KiB (x8)
  L2 Unified 512 KiB (x2)
Load Average: 2.46, 2.66, 2.42
---------------------------------------------------------------------
Benchmark                           Time             CPU   Iterations
---------------------------------------------------------------------
MLIRPolybench2mm/scalar/0       0.535 ms        0.535 ms         1307
MLIRPolybench2mm/scalar/1        12.0 ms         12.0 ms           58
MLIRPolybench2mm/scalar/2         653 ms          651 ms            1
MLIRPolybench2mm/autovec/0      0.108 ms        0.108 ms         6476
MLIRPolybench2mm/autovec/1       2.42 ms         2.41 ms          290
MLIRPolybench2mm/autovec/2        185 ms          185 ms            4
------------------------------------------------
autovec 2mm-mini PASS
autovec 3mm-mini PASS
autovec adi-mini PASS
autovec atax-mini PASS
autovec bicg-mini (s) PASS
autovec bicg-mini (q) PASS
autovec cholesky-mini PASS
autovec correlation-mini PASS
autovec covariance-mini PASS
autovec deriche-mini PASS
autovec doitgen-mini PASS
autovec durbin-mini PASS
autovec fdtd-2d-mini (ex) PASS
autovec fdtd-2d-mini (ey) PASS
autovec fdtd-2d-mini (hz) PASS
autovec floyd-warshall-mini PASS
autovec gemm-mini PASS
autovec gemver-mini PASS
autovec gesummv-mini PASS
autovec gramschmidt-mini (R) PASS
autovec gramschmidt-mini (Q) PASS
autovec heat-3d-mini PASS
autovec jacobi-1d-mini PASS
autovec jacobi-2d-mini PASS
autovec lu-mini PASS
autovec ludcmp-mini PASS
autovec mvt-mini (x1) PASS
autovec mvt-mini (x2) PASS
autovec nussinov-mini PASS
autovec seidel-2d-mini PASS
autovec symm-mini PASS
autovec syr2k-mini PASS
autovec syrk-mini PASS
autovec trisolv-mini PASS
autovec trmm-mini PASS
------------------------------------------------
```

xlinsist commented on benchmarks/Vectorization/polybench/polybench_mlir_gen.py:
This script provided here is thoughtful~

xlinsist commented on benchmarks/Vectorization/polybench/README.md:
````diff
@@ -0,0 +1,155 @@
+```
+
+3. Run the benchmark:
+
+```bash
+./bin/vectorization-polybench-benchmark --benchmark_filter=.*/.*/[12]
+```
````
1. This part can be removed and put together with "## Running the Benchmark" below.
2. Better to add explanations that `--benchmark_filter=.*/.*/[12]` is a command line option for google benchmark, and this regular expression will filter out all benchmarks for small and medium datasets, since the size indexes of these two datasets are 1 and 2 respectively.

JuniMay replied:
Updated README and added explanations for the filters.

xlinsist commented:
> The filter can also be used to run only certain Polybench cases.

Commands like `./bin/vectorization-polybench-benchmark --benchmark_filter="MLIRPolybench2mm/.*/[012]"` can be added to the README as an example usage.

xlinsist commented:
Though some parts of the code require modification and refactoring, this PR is overall well-completed. It also includes verification scripts and test results, which I have reviewed. Well done!

Out of curiosity, since polybench includes many cases, have you sought assistances from hand-written automated/semi-automated scripts, to help improve your efficiency?"

JuniMay commented:
Thanks, and I'm working on the modification of the required parts.

And yes, I tried to automate the porting process from the beginning of this PR. The script provided under the `polybench` directory is used to generate all MLIR codes with Polygeist, and to verify the results conveniently. I also tried using LLMs to modify MLIR files and generate the benchmarking code, but it turned out to be a failure (I tried to prompt engineer Qwen, 4o, DeepSeek, etc., but failed to generate expected code).

There are some repetitive parts that can be simplified with macros, like `_scalar` and `_autovec` method names. However, considering the different kernel signatures and that we might want to optimize a single benchmark case with specialized pipelines (which requires flexibility over simplicity) I kept them the same as the current implementation.

JuniMay commented on benchmarks/Vectorization/polybench/CMakeLists.txt:
```diff
@@ -0,0 +1,100 @@
+    VERBATIM
+  )
+
+  # [Step 4] Add more compilation pipeline steps here if needed
```
Yes, that makes sense, already removed.

JuniMay commented on benchmarks/Vectorization/polybench/Utils.hpp:
```diff
@@ -0,0 +1,72 @@
+
+} // namespace polybench
+
+#endif // POLYBENCH_UTILS_HPP
```
Fixed.

JuniMay commented on benchmarks/Vectorization/polybench/Main.cpp:
```diff
@@ -0,0 +1,225 @@
+    return 4;
+  } else {
+    return -1;
+  }
```
Fixed, moved them into `Utils.hpp`.

JuniMay commented on benchmarks/Vectorization/polybench/Main.cpp:
```diff
@@ -0,0 +1,225 @@
+    generateResultMLIRPolybenchSyr2k(size_id);
+    generateResultMLIRPolybenchSyrk(size_id);
+    generateResultMLIRPolybenchTrisolv(size_id);
+    generateResultMLIRPolybenchTrmm(size_id);
```
All the lengthy function declarations and helper macros are now inside `Utils.hpp` to simplify `Main.cpp`.

xlinsist commented:
> The script provided under the polybench directory is used to generate all MLIR codes with Polygeist, and to verify the results conveniently. I also tried using LLMs to modify MLIR files and generate the benchmarking code, but it turned out to be a failure (I tried to prompt engineer Qwen, 4o, DeepSeek, etc., but failed to generate expected code).
I see.

> considering the different kernel signatures and that we might want to optimize a single benchmark case with specialized pipelines

It makes sense. We also need to beware of the trap of oversimplifying code.

xlinsist commented:
LGTM. Thanks!
